File: content/snippets/cpp/design_patterns/mediator.cpp
#include <iostream>
#include <memory>

class Mediator {
public:
    virtual ~Mediator() = default;
    virtual void notify(void* sender, const std::string& event) = 0;
};

class User {
protected:
    std::shared_ptr<Mediator> mediator;
    std::string name;
public:
    User(std::shared_ptr<Mediator> m, const std::string& n) : mediator(m), name(n) {}
    virtual ~User() = default;
    virtual void send(const std::string& msg) = 0;
    virtual void receive(const std::string& msg) = 0;
};

class ChatUser : public User {
public:
    using User::User;
    
    void send(const std::string& msg) override {
        std::cout << name << " sends: " << msg << std::endl;
        mediator->notify(this, msg);
    }
    
    void receive(const std::string& msg) override {
        std::cout << name << " receives: " << msg << std::endl;
    }
};

class ChatRoom : public Mediator {
private:
    std::shared_ptr<ChatUser> user1, user2;
public:
    void addUser(std::shared_ptr<ChatUser> u1, std::shared_ptr<ChatUser> u2) {
        user1 = u1;
        user2 = u2;
    }
    
    void notify(void* sender, const std::string& event) override {
        if (sender == user1.get()) {
            user2->receive(event);
        } else {
            user1->receive(event);
        }
    }
};

int main() {
    auto room = std::make_shared<ChatRoom>();
    auto u1 = std::make_shared<ChatUser>(room, "Alice");
    auto u2 = std::make_shared<ChatUser>(room, "Bob");
    
    room->addUser(u1, u2);
    
    u1->send("Hello!");
    u2->send("Hi there!");
}

File: content/snippets/python/design_patterns/mediator.py
from abc import ABC, abstractmethod

class Mediator(ABC):
    @abstractmethod
    def notify(self, sender: "User", event: str) -> None:
        pass

class User:
    def __init__(self, mediator: Mediator, name: str):
        self.mediator = mediator
        self.name = name
    
    def send(self, msg: str) -> None:
        print(f"{self.name} sends: {msg}")
        self.mediator.notify(self, msg)
    
    def receive(self, msg: str) -> None:
        print(f"{self.name} receives: {msg}")

class ChatRoom(Mediator):
    def __init__(self):
        self.user1: User = None
        self.user2: User = None
    
    def add_users(self, u1: User, u2: User) -> None:
        self.user1 = u1
        self.user2 = u2
    
    def notify(self, sender: User, event: str) -> None:
        if sender == self.user1:
            self.user2.receive(event)
        else:
            self.user1.receive(event)

room = ChatRoom()
u1 = User(room, "Alice")
u2 = User(room, "Bob")

room.add_users(u1, u2)

u1.send("Hello!")
u2.send("Hi there!")
